scripts: add script to extract top-ten bug-bashers

cfriedt opened the pull request:
This script allows us to programmatically query bug-bashers within a user-supplied time-window.

For example, we held a "Bug Bash Week" August 1-7, 2021 (it was announced a week early though). The output of the script prints the "top ten" bug bashers in tab-separated columns in descending order. The first column is the number of bugs squashed and the second column is the github user id.

```
./bug_bash.py -t ~/.ghtoken -b 2021-07-26 -e 2021-08-07
3       LingaoM
2       de-nordic
2       cvinayak
1       lairdjm
1       anangl
1       jfischer-no
1       edmont
1       mbolivar-nordic
1       nandojve
1       nordic-krch
```

cfriedt committed:
scripts: add script to extract top-ten bug-bashers

This script allows us to programmatically query bug-bashers within
a user-supplied time-window.

For example, we held a "Bug Bash Week" August 1-7, 2021 (it was
announced a week early though). The output of the script prints
the "top ten" bug bashers in tab-separated columns in descending
order. The first column is the number of bugs squashed and the
second column is the github user id.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

mbolivar-nordic commented on scripts/release/bug_bash.py:
```diff
@@ -0,0 +1,153 @@
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-t', '--token', dest='token',
+                        help='GitHub token', metavar='token', required=True)
```
IMO it's bad practice to encourage leaving credentials in .bash_history. I would rather see something like this:

https://github.com/nrfconnect/sdk-nrf/blob/0fc8888738cbdc810d43d85402255f9fe5421b95/scripts/get_pulls_in_range.py#L150

I just noticed that script is missing a copyright and license, so I added one here:

https://github.com/nrfconnect/sdk-nrf/pull/5320

Once that is merged you should just be able to copy/paste into here if you want.

cfriedt replied:
Good call - I may do something similar where there is an option to specify a credentials file and then fallback to `GITHUB_TOKEN` / error if unset.

mbolivar-nordic reviewed (changes requested):
Just one credential objection; otherwise seems fine.

cfriedt commented:
* modified args to read Github token from a specified file or the `GITHUB_TOKEN` environment variable so that sensitive data was not left in bash history
* reduced indentation a bit
* added `PyGithub` to `scripts/requirements-extras.txt`

mbolivar-nordic commented on scripts/release/bug_bash.py:
```diff
@@ -0,0 +1,187 @@
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-a', '--all', dest='all',
+                        help='Show all bugs squashed', action='store_true', default=False)
```
```suggestion
                        help='Show all bugs squashed', action='store_true')
```

Just FYI, this is redundant with store_true.

mbolivar-nordic commented on scripts/release/bug_bash.py:
```diff
@@ -0,0 +1,187 @@
+    parser.add_argument('-a', '--all', dest='all',
+                        help='Show all bugs squashed', action='store_true', default=False)
+    parser.add_argument('-t', '--token', dest='tokenfile',
+                        help='File containing GitHub token', metavar='FILE', default=None)
```
```suggestion
                        help='File containing GitHub token', metavar='FILE')
```

This is also redundant.

mbolivar-nordic commented on scripts/release/bug_bash.py:
```diff
@@ -0,0 +1,187 @@
+
+    if args.end < args.begin:
+        raise ValueError(
+            'end date {} is before begin date {}'.format(args.end, args.begin))
```
Just a general note that Zephyr requires python 3.6+, so you can use f-strings if you want.

mbolivar-nordic reviewed (approved):
Nonblocking comments

cfriedt commented:
* remove redundant defaults from arguments
